pgml-cms/blog/semantic-search-in-postgres-in-15-minutes.md

```diff
@@ -0,0 +1,467 @@
+
+The output size of the vector varies per model, and in _mxbai-embed-large-v1_ outputs vectors with 1024 dimensions: each vector contains 1024 floating point numbers. 
+
+The vector this model outputs is not random. It is designed to capture the semantic meaning of the text. What this really means, is that sentences which are closer together in meaning will be closer together in vector space. 
+
+Let’s look at a more simple example. Let's assume we have a model called _simple-embedding-model_, and it outputs vectors with only 2 dimensions. Let’s embed the following three phrases: "I like Postgres", "I like SQL" and "Rust is the best": 
+
+!!! generic
+
+!!! code_block
+
+```postgresql
+SELECT pgml.embed('simple-embedding-model', 'I like Postgres') AS embedding;
+
+SELECT pgml.embed('simple-embedding-model', 'I like SQL') AS embedding;
+
+SELECT pgml.embed('simple-embedding-model', 'Rust is the best') AS embedding;
+```
+
+!!!
+
+!!! results
+
+```text
+embedding for 'I like Postgres'
+---------
+[0.1, 0.2]
+
+embedding for 'I like SQL'
+---------
+[0.12, 0.25]
+
+embedding for 'Rust is the best'
+---------
+[-0.8, -0.9]
+```
+
+!!!
+
+!!!
+
+You'll notice how similar the vectors produced by the text "I like Postgres" and "I like SQL" are compared to "Rust is the best".  This is a artificial example, but the same idea holds true when translating to real models like _mixedbread-ai/mxbai-embed-large-v1_. 
+
+## What does it mean to be "close"?
+
+We can use the idea that text that is more similar in meaning will be closer together in the vector space to build our semantic search engine. 
+
+For instance let’s say that we have the following documents:
+
+| Document ID | Document text |
+-----|----------|
+| 1 | The pgml.transform function is a PostgreSQL function for calling LLMs in the database. |
+| 2 | I think tomatoes are incredible on burgers. |
+
+
+and a user is looking for the answer to the question: "What is the pgml.transform function?". If we embed the search query and all of the documents using a model like _mixedbread-ai/mxbai-embed-large-v1_, we can compare the query embedding to all of the document embeddings, and select the document that has the closest embedding in vector space, and therefore in meaning, to the to the answer. 
+
+These are big embeddings, so we can’t simply estimate which one is closest. So, how do we actually measure the similarity (distance) between different vectors?
+
+_pgvector_ as of this writing supports four different measurements of vector similarity:
+
+- L2 distance
+- (negative) inner product
+- cosine distance
+- L1 distance
+
+For most use cases we recommend using the cosine distance as defined by the formula:
```

Comment:
You may be able to shorten this section with a link to our docs on distance functions.

Reply:
I absolutely can, but do we want to? I was thinking this would post would give enough of an intro that after going through
it they would feel confident enough to reproduce this on their own test project, or rather, use this as a guide and copy the steps we did without too much hand waving. If the length of this section takes away from that goal I will trim it down!